src/cr/cube/dimension.py

```diff
@@ -19,124 +21,125 @@ class Dimension(object):
-        return refs.get('description')
+        """str description of this dimension."""
+        description = self._dimension_dict['references'].get('description')
+        return description if description else ''
+
+    @lazyproperty
+    def dimension_type(self):
+        """str representing type of this cube dimension."""
+        # ---all this logic really belongs in the Dimensions collection
+        # ---object, which is where it will move to once that's implemented
+
+        def next_dim_is_mr_cat():
+            """True if subsequent dimension is an MR_CAT dimension."""
+            if not self._next_dimension_dict:
+                return False
+
+            categories = self._next_dimension_dict['type'].get('categories')
+            if not categories:
+                return False
+
+            return (
+                [category.get('id') for category in categories] == [1, 0, -1]
+            )
+
+        type_dict = self._dimension_dict['type']
+        type_class = type_dict.get('class')
+
+        if not type_class:
+            # ---numeric and text are like this---
+            return type_dict['subtype']['class']
+
+        if type_class == 'enum':
+            if 'subreferences' in self._dimension_dict['references']:
+                return (
+                    'multiple_response' if next_dim_is_mr_cat()
+                    else 'categorical_array'
+                )
+            if 'subtype' in type_dict:
+                # ---datetime is like this (enum without subreferences)---
+                return type_dict['subtype']['class']
+
+        return type_class
 
     @memoize
-    def elements(self, include_missing=False):
-        """Get elements of the crunch Dimension.
+    def element_indices(self, include_missing):
+        """Return tuple of int element idxs for this dimension.
 
-        For categorical variables, the elements are represented by categories
-        internally. For other variable types, actual 'elements' of the
-        Crunch Cube JSON response are returned.
+        *include_missing* determines whether missing elements are included or
+        only valid element index values are returned.
         """
-        if include_missing:
-            return self._elements
-
-        return [
-            el for (i, el) in enumerate(self._elements)
-            if i not in self.invalid_indices
-        ]
+        return (
+            self._all_elements.element_idxs if include_missing else
+            self._valid_elements.element_idxs
+        )
 
-    @lazyproperty
-    def elements_by_id(self):
-        r = {}
-        for i, el in enumerate(self._elements):
-            el['index'] = i
-            r[el['id']] = el
-        return r
+    @memoize
+    def elements(self, include_missing=False):
+        """_Elements object providing access to elements of this dimension."""
+        return (
+            self._all_elements if include_missing else self._valid_elements
+        )
```

Comment:
One of the big items in this PR is extracting `_Elements` and `_Element` (and their object hierarchy siblings) as full-fledged objects. This removes all the widely distributed dict manipulation that is hard to follow and very error prone. Using dicts instead of real objects for jobs like these is an example of the "Primitive Obsession" anti-pattern, and this refactoring provides a good example of how to eliminate that and the benefits in clarity, conciseness, and testability that can bring.

Reply:
❤️